### PR TITLE
Added support for Stream encryption/file encryption using xchacha20-poly1305

### PR DESCRIPTION
**Pull request by dkulic**
Signed-off-by: Darko Kulic <darko.kulic@evernym.com>

### CODE REVIEW

**Comment by dkulic** [comment]
This is resubmitted PR #202

**Comment by nvesely** [comment]
> Signed-off-by: Darko Kulic darko.kulic@evernym.com

Did you mean to sign this with your GPG key (which is `git commit -S`) or are you using `git commit -s` because you wish to explicitly say that you are making your contributions available under the same license as the project?

>        -s, --signoff
>           Add Signed-off-by line by the committer at the end of the commit
>           log message. The meaning of a signoff depends on the project, but
>           it typically certifies that committer has the rights to submit this
>           work under the same license and agrees to a Developer Certificate
>           of Origin (see http://developercertificate.org/ for more
>           information).

>       -S[<keyid>], --gpg-sign[=<keyid>]
>           GPG-sign commits. The keyid argument is optional and defaults to
>           the committer identity; if specified, it must be stuck to the
>           option without a space.

**Comment by marinintim** [comment]
The Travis build is failed for unrelated reasons (docs in Serde).

**Comment by marinintim** [comment]
@nvesely: what do we need to merge this PR, an answer from @dkulic?

**Comment by Dylan-DPC** [comment]
@marinintim will rerun the CI process. Sorry for the delay :slightly_smiling_face:

**Comment by dkulic** [comment]
I wanted   -s, --signoff
      Add Signed-off-by line by the committer at the end of the commit
      log message. The meaning of a signoff depends on the project, but
      it typically certifies that committer has the rights to submit this
      work under the same license and agrees to a Developer Certificate
      of Origin (see http://developercertificate.org/ for more
      information).

**Comment by Dylan-DPC** [comment]
@sodiumoxide/owners can we get a review of this? would prefer one review from rust perspective and one from a cryptographic perspective.

**Comment by nvesely** [comment]
> @nvesely: what do we need to merge this PR, an answer from @dkulic?

No, I was just curious. When I first started using git, I thought I was signing my messages when I was really "signing-off," and wanted to make sure dkulic wasn't making the same error.

> @sodiumoxide/owners can we get a review of this? would prefer one review from rust perspective and one from a cryptographic perspective.

I've been traveling a lot, but think I can give this a second review later today.

On another note, I would like to discuss this review process idea in another ticket or on Gitter at a later time. I think the task of deciding what is a useful and correct PR should involve a more nuanced process, and together we could all document an outline of a more complete process.

**Comment by dnaq** [comment]
@tom-phinney I haven’t had the time to review the C code yet, do you know if the authentication for each message depends on the previous message as well. That is, is it possible to withhold or reorder each individual message without the receiver noticing until receiving and authenticating the final message?

When it comes to the code I see no real need for using a macro, unless there are other primitives implementing the same api.

**Comment by Dylan-DPC** [comment]
> On another note, I would like to discuss this review process idea in another ticket or on Gitter at a later time. I think the task of deciding what is a useful and correct PR should involve a more nuanced process, and together we could all document an outline of a more complete process.

@nvesely i feel we could end up overengineering the review process. I feel it is fine the way it is, as long as at least 2 people review the code from a rust perspective and 1-2 from the crypto perspective.

**Comment by samscott89** [comment]
@dnaq : The [libsodium docs](https://download.libsodium.org/doc/secret-key_cryptography/secretstream.html) claim re-ordering isn't possible:
> Messages cannot be truncated, removed, reordered, duplicated or modified without this being detected by the decryption functions.

I've haven't checked the C code to verify this. It also might be interesting to check whether it satisfies the notions put forward in [2017/1191](https://eprint.iacr.org/2017/1191).

**Comment by Tom-Phinney** [comment]
@samscott89 crypto_secretstream_xchacha20poly1305.h defines the tags; the Rust code merely interfaces correctly to the C code, AFAI can tell. Changes to the Rust API of the extent you suggest would not result in an honest mirroring of the libsodium secretstream API in Rust.

**Comment by Tom-Phinney** [comment]
> @dnaq: When it comes to the code I see no real need for using a macro, unless there are other primitives implementing the same api.

The macros simply provide a way to mirror in Rust what is currently supported in C in libsodium. It's not how I would design the API, but it does have the advantage that even the most distracted programmer probably can't screw up the libsodium invocations.

**Comment by dnaq** [comment]
I believe that sodiumoxide should expose a high-level rust api where the
possibility of invoking it wrongly should be as small as possible, so I
agree on the comment that we shouldn’t expose the individual tags. Instead
we should have methods for push_message and finalize, where finalize
consumes the object (takes mut self and doesn’t return it again), the names
could be different, but should be chosen to be consistent with the other
APIs. Rekey should also be exposed as a method.

It’s never been a goal of sodiumoxide to mirror the C APIs, the goal has
always been to expose them in a rustish (rusty?) way.

**Comment by samscott89** [comment]
@Tom-Phinney
> crypto_secretstream_xchacha20poly1305.h defines the tags; the Rust code merely interfaces correctly to the C code, AFAI can tell. Changes to the Rust API of the extent you suggest would not result in an honest mirroring of the libsodium secretstream API in Rust.

I would disagree. Perhaps there are less significant changes which would offer similar guarantees. But I think there's an opportunity to use Rust's features to provide an interface which is actually closer to the intended functionality. Same as how we use newtypes for nonces, etc. `TAG_FINAL` is supposed to end the stream, but currently you can continue to send messages afterwards.

**Comment by Tom-Phinney** [comment]
I guess I misunderstood the goals of the sodiumoxide crate, perhaps because I conflated them with the goal of providing an interface to the C sodium chloride (NaCl) library. If there is no need to maintain API compatibility, then I agree that a more oxidized (i.e., rusty) version of the secretstream functionality is appropriate.

Of course such changes/evolution in the API implies that any existing documentation for the C secretstream API will need substantial reworking to match the Rusty API, followed by review and testing of the documentation. Thus it's likely to be some time before this PR is complete enough to merge.

**Comment by jedisct1** [comment]
> `TAG_FINAL` is supposed to end the stream, but currently you can continue to send messages afterwards.

Semantically, it ends the stream. It is totally fine to consume `self` here.

The main reason the C library performs a rekey on `TAG_FINAL` is to mitigate security implications on misuse. `crypto_generic_hash_final()` was supposed to end the stream as well, but I saw people call it multiple times no matter what.

Rust ownership system prevents this.

File: src/crypto/secretstream/secretstream_macros.rs [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            c.set_len(clen as usize);
+        }
+        c
+    }
+    
+    /// function verifies that c (a sequence of bytes) contains a valid ciphertext and authentication tag for the given state state and optional authenticated data ad of length adlen bytes.
+    /// If the ciphertext appears to be invalid, the function returns Err.
+    /// If the authentication tag appears to be correct, the decrypted message is returned with tag.
+    /// Applications will typically call this function in a loop, until a message with the crypto_secretstream_xchacha20poly1305_TAG_FINAL tag is found.
+    pub fn pull(&mut self, c: &[u8], ad: Option<&[u8]>) -> Result<(Vec<u8>, u8),()> {
```
Comment (nvesely) [inline src/crypto/secretstream/secretstream_macros.rs]: You should return `Err(())` if `c.len() < ABYTES` or if `mlen < MESSAGEBYTES_MAX`. If you test them in order then underflow for the subtraction won't be a worry.

**Comment by samscott89** [comment]
@jedisct1 Thanks for the explanation/clarification (and also for explaining where `TAG_FINAL` is defined in the other thread).

File: src/crypto/secretstream/secretstream_macros.rs [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    secret Key(KEYBYTES);
+}
+
+/// `gen_key()` randomly generates a key for symmetric encryption
+///
+/// THREAD SAFETY: `gen_key()` is thread-safe provided that you have
+/// called `sodiumoxide::init()` once before using any other function
+/// from sodiumoxide.
+pub fn gen_key() -> Key {
+    let mut key = [0; KEYBYTES];
```
[the remark this answers is not in the document]
Comment (samscott89) [reply src/crypto/secretstream/secretstream_macros.rs]: I was wondering about this too. Playing around with godbolt a little, I think the compiler (on `--release`) will generally optimize away the difference. But if for whatever reason the libsodium call doesn't set the memory, it's bad news.

(Rip)grepping through the repo, generally simple arrays are initialized to 0s and anything more complex like structs use `mem::uninitialized`. I think this is a good practice to follow.

**Comment by samscott89** [comment]
Okay, the `push_*` API stuff seems like it's heading in a good direction. The `pull_*` API may be a little harder to write cleanly. It's bound to be the most complex part of the API, due to the streaming/async-like nature of it.

A couple of suggestions: implement `Read` or `Iterator` for it? Or simply `pull` just returns `Result<Option<Vec<u8>>, ()>` ? This would handle the final case: simply returns `None` afterwards. The library user can then easily turn this into an iterator or whatever. But then I'm not sure whether `TAG_PUSH` would have any semantics?

Maybe even `Iterator<Item = Iterator<Item = Vec<u8>>>`? The inner iterator returns `Some(vec)` for each message received until hitting `TAG_PUSH`, and the outer iterator returns inner iterators until hitting `TAG_FINAL`?

**Comment by dnaq** [comment]
Couldn’t we make pull take mut self and return an enum with the result, and itself. E.g. (pseudocode)
‘
enum Result {
    Message(mut

**Comment by nvesely** [comment]
> A couple of suggestions: implement Read or Iterator for it?

What's "it"? The `State` struct? I'm not sure how you're envisioning this working with a separate stream of incoming ciphertexts? Can you sketch this out a bit more for me.

---

Here's my idea expanding on the suggestion of splitting state into `ADecryptor` and `AEncryptor`. The `ADecryptor` struct contains a `state` and `finalized` field and implementes five methods: `init`, `rekey`, `adecrypt`, `is_finalized`, and `is_not_finalized`. `adecrypt` returns an `Err(())` if `finalized` is `true`, but otherwise works almost like the version in this PR (only other change I can think of is returning a `Tag` enum instead of `u8`).

```rust
pub struct ADecryptor {
    state: secretstream_state,
    finalized: bool,
}
    
let header = socket.read();
let mut decryptor = Decryptor::init(&header, &key);
while decryptor.is_not_finalized() {
    let (c, ad) = socket.read();
    let m = decryptor.auth_decrypt();
    // Do something with `m` which is either an `Ok(Vec<u8>, Tag)` or `Err(())`
}
```

A corresponding `Encryptor` would just be a simple alias for `secretstream_state`. The alias would have an `init` and `rekey` methods, and the various `aencrypt_` methods corresponding to the 4 tags. Each of these methods would take a `&mut self` except for `aencrypt_finalize`, which would take `self`.

I think the separation of `State` into `Decryptor` and `Encryptor` objects would make both for a clearer, easier-to-use API, and prevent errors from potentially calling `pull` on `State` initialized via `init_push` or vice-versa.

---

> But then I'm not sure whether TAG_PUSH would have any semantics?

I think the user is going to have to decide how to handle `TAG_PUSH`. It could be used in different ways in different applications.

---

> There should be no need to expose the rekey tag.

It is true that when an authenticated ciphertext containing the rekey tag is received, the state is automatically updated accordingly by `crypto_secretstream_xchacha20poly1305_push`. However, I can imagine a use-case whereupon receiving a rekey tag additional changes outside of mutating the state objects need to be carried out. Further, having `pull`/`adecrypt` be able to return a `Ok((Vec<u8>, Tag))` is more simple than having to specify `Ok((Vec<u8>, Option<Tag>))` where we return `None` instead of `Tag::Rekey` (of course there maybe other ways to reduce this complexity).

**Comment by nvesely** [comment]
It would also be possible to use a single struct to keep the API a little closer to upstream, while still receiving the safety benefits provided by the use of separate structs above. E.g.,

```rust
pub struct State {
    state: secretstream_state,
    finalized: Option<bool>
}
```

`init_push` would set finalize to `None`, and `init_pull` would set it to `Some(false)`. Then `pull` and `is_{not_}finalized` will return `Err(())` if `finalized.is_none()`, while the `push_*` methods will return `Err(())` if `finalized.is_some()`.

I think this provides a good balance between not diverging too far from the upstream API, while making it as difficult as possible for the programmer to screw anything up.

**Comment by dkulic** [comment]
Thanks all, I will try in next several days to find some time to make improvements requested.

**Comment by nvesely** [comment]
See my branch `nvesely:secretstream` or [specifically the last WIP commit](https://github.com/nvesely/sodiumoxide/commit/0fb9fa6de1926696fb3de18e158baccc4e45e079). I ran into some unexpected errors implementing the `Tag` enum. Maybe @sodiumoxide/owners can take a look.

Casting a `Tag` to a `u8` is easily done using `as`, whereas casting a `u8` to a `Tag` should return a `Result<Tag, ()>`. The problem with my function which does this, `custom_try_from`, is that it always matches the first arm of the `match` expression, even if the `u8`s are not the same value. I thought I might be making a small syntactical error, but I don't see how this is any different from the first example from https://doc.rust-lang.org/reference/expressions/match-expr.html. I have added some `println` statements to try to debug. Try playing around with the match expression and testing using `cargo test test_push_pull_with_rekey -- --nocapture`.

**Comment by Dylan-DPC** [comment]
@nvesely maybe you can use TryFrom Trait and feature-gate it till it is stabilised?

**Comment by nvesely** [comment]
@Dylan-DPC Yes, but that's a separate issue.

**Comment by samscott89** [comment]
@nvesely The warning messages in this case show what's happening:
```
warning: unreachable pattern
  --> src/crypto/secretstream/secretstream_macros.rs:64:13
   |
62 |               push => Ok(Tag::Push),
   |               ---- matches any value
63 |               message => Ok(Tag::Message),
64 |               _ => Err(())
   |               ^ unreachable pattern
```
 In this case, you're trying to match against a non-constant expression, so the variables `push` and `message` are actually shadowing the variables you created beforehand. `push` just captures any value.

We could consider using `enum_primitive_derive` as suggested here:  https://github.com/rust-lang/rust/issues/44266#issuecomment-326774902. Or recreating something similar.

**Comment by nvesely** [comment]
Thanks, I missed that warning. Got it working now with `enum_primitive_derive`, which I like.

File: src/crypto/secretstream/secretstream_macros.rs [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+/// to process a message as a sequence of multiple chunks.
+pub struct State($state_name);
+
+/// Initializes a `state` using the `key` and an internal, automatically generated initialization vector.
+/// It then stores the stream header into header (crypto_secretstream_xchacha20poly1305_HEADERBYTES bytes).
+pub fn init_push(key: &Key) -> (State, [u8; HEADERBYTES]) {
+    unsafe {
+        let mut state: $state_name = mem::uninitialized();
+        let mut header: [u8; HEADERBYTES] = mem::uninitialized();
+        $init_push_name(&mut state, header.as_mut_ptr(), key.0.as_ptr());
```
[the remark this answers is not in the document]
Comment (dkulic) [reply src/crypto/secretstream/secretstream_macros.rs]: I checked the c code, it always returns 0.

File: src/crypto/secretstream/secretstream_macros.rs [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        }
+        Ok(State(state))
+    }
+}
+
+impl State {
+
+    /// encrypts a message `m` using the `state` and the `tag`.
+    /// Additional data ad of length adlen can be included in the computation of the authentication tag. If no additional data is required, ad can be None.
+    pub fn push(&mut self, m: &[u8], ad: Option<&[u8]>, tag: u8) -> Vec<u8> {
```
[the remark this answers is not in the document]
Comment (dkulic) [reply src/crypto/secretstream/secretstream_macros.rs]: OK

File: src/crypto/secretstream/secretstream_macros.rs [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        let (ad_p, ad_len) = ad.map(|ad| (ad.as_ptr(), ad.len() as c_ulonglong)).unwrap_or((0 as *const _, 0));
+        let mut c = Vec::with_capacity(m.len() + ABYTES);
+        let mut clen = c.len() as c_ulonglong;
+        
+        unsafe {
+            $push_name(&mut self.0,
```
[the remark this answers is not in the document]
Comment (dkulic) [reply src/crypto/secretstream/secretstream_macros.rs]: Sure. It always return 0, but we would already have result in response type so it is free.

File: src/crypto/secretstream/mod.rs [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+//!
+//! This API can be used to securely send an ordered sequence of messages to a peer.
+//! Since the length of the stream is not limited, it can also be used to encrypt files
+//! regardless of their size.
+//!
+//! It transparently generates nonces and automatically handles key rotation.
+//!
+//! The crypto_secretstream_*() API was introduced in libsodium 1.0.14.
+//!
+//! # Example (stream encryption)
```
[the remark this answers is not in the document]
Comment (dkulic) [reply src/crypto/secretstream/mod.rs]: Good catch.

File: src/crypto/secretstream/secretstream_macros.rs [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+use randombytes::randombytes_into;
+use std::mem;
+
+/// Number of bytes in a `Key`.
+pub const KEYBYTES: usize = $keybytes as usize;
+
+/// Number of bytes in a `Header`.
+pub const HEADERBYTES: usize = $headerbytes as usize;
+
+/// Number of added bytes
```
[the remark this answers is not in the document]
Comment (dkulic) [reply src/crypto/secretstream/secretstream_macros.rs]: Sure

**Comment by jedisct1** [comment]
Although I never had the opportunity to use sodiumoxide yet, I just wanted to say a big *thank you* to you people who have been building these fantastic bindings, and designed very nice, idiomatic APIs.

`secretstream` was probably the trickiest piece of the puzzle. As a similar operation is eventually going to be added to libhydrogen, your input would be very valuable.

How would you have designed the `secretstream` C API so that writing Rust bindings would have been easier? What did you find confusing about it, both in the API and the documentation? Anything you like/don't like about it?

Thanks!

**Comment by Dylan-DPC** [comment]
Thanks @jedisct1. I think it is better if we discuss this in a separate issue.

File: src/crypto/secretstream/xchacha20poly1305.rs [hunk 8]
```diff
@@ -0,0 +1,160 @@
+//! `crypto_stream_chacha20` (Chacha20)
```
[the remark this answers is not in the document]
Comment (nvesely) [reply src/crypto/secretstream/xchacha20poly1305.rs]: Fixed in https://github.com/sodiumoxide/sodiumoxide/pull/222/commits/4133121eb6294184f6f0e22948bc294989fb9dc5.

File: src/crypto/secretstream/secretstream_macros.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (nvesely) [reply src/crypto/secretstream/secretstream_macros.rs]: It imports the function `crypto_secretstream_xchacha20poly1305_messagebytes_max`, which computes the value. I have code I can use to add a commit to this PR.

[the remark this answers is not in the document]
Comment (dkulic) [reply src/crypto/secretstream/secretstream_macros.rs]: Sure, create PR. That length is checked in C code also, but I think it does not return the value but aborts the program (did not spend a lot of time checking).

File: src/crypto/secretstream/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (nvesely) [reply src/crypto/secretstream/mod.rs]: Addressed in https://github.com/sodiumoxide/sodiumoxide/pull/222/commits/4133121eb6294184f6f0e22948bc294989fb9dc5, though note contains spelling error and could also have improved content. I have already written something else I can add.

File: src/crypto/secretstream/mod.rs [hunk 9]
```diff
@@ -0,0 +1,80 @@
+//! Stream encryption/file encryption
```
[the remark this answers is not in the document]
Comment (nvesely) [reply src/crypto/secretstream/mod.rs]: Better to beg forgiveness, than ask permission I guess... let's go with that...

File: src/crypto/secretstream/secretstream_macros.rs [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+
+#[cfg(not(feature = "std"))] use prelude::*;
+use libc::c_ulonglong;
+use randombytes::randombytes_into;
+use std::mem;
+
+/// Number of bytes in a `Key`.
+pub const KEYBYTES: usize = $keybytes as usize;
+
+/// Number of bytes in a `Header`.
```
[the remark this answers is not in the document]
Comment (nvesely) [reply src/crypto/secretstream/secretstream_macros.rs]: Fixed in https://github.com/sodiumoxide/sodiumoxide/pull/222/commits/4133121eb6294184f6f0e22948bc294989fb9dc5.

File: src/crypto/secretstream/secretstream_macros.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (nvesely) [reply src/crypto/secretstream/secretstream_macros.rs]: Fixed in https://github.com/sodiumoxide/sodiumoxide/pull/222/commits/4133121eb6294184f6f0e22948bc294989fb9dc5.

File: src/crypto/secretstream/secretstream_macros.rs [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+pub const HEADERBYTES: usize = $headerbytes as usize;
+
+/// Number of added bytes
+pub const ABYTES: usize = $abytes as usize;
+
+/// Tag message
```
[the remark this answers is not in the document]
Comment (nvesely) [reply src/crypto/secretstream/secretstream_macros.rs]: Fixed in https://github.com/sodiumoxide/sodiumoxide/pull/222/commits/4133121eb6294184f6f0e22948bc294989fb9dc5.

File: src/crypto/secretstream/secretstream_macros.rs [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+pub const TAG_MESSAGE: u8 = $tag_message as u8;
+
+/// Tag push
```
[the remark this answers is not in the document]
Comment (nvesely) [reply src/crypto/secretstream/secretstream_macros.rs]: Fixed in https://github.com/sodiumoxide/sodiumoxide/pull/222/commits/4133121eb6294184f6f0e22948bc294989fb9dc5.

File: src/crypto/secretstream/secretstream_macros.rs [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+pub const TAG_PUSH: u8 = $tag_push as u8;
+
+/// Tag rekey
+pub const TAG_REKEY: u8 = $tag_rekey as u8;
+
+/// Tag final
```
[the remark this answers is not in the document]
Comment (nvesely) [reply src/crypto/secretstream/secretstream_macros.rs]: Fixed in https://github.com/sodiumoxide/sodiumoxide/pull/222/commits/4133121eb6294184f6f0e22948bc294989fb9dc5.

File: src/crypto/secretstream/secretstream_macros.rs [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+/// Number of added bytes
+pub const ABYTES: usize = $abytes as usize;
+
+/// Tag message
+pub const TAG_MESSAGE: u8 = $tag_message as u8;
+
+/// Tag push
+pub const TAG_PUSH: u8 = $tag_push as u8;
+
+/// Tag rekey
```
[the remark this answers is not in the document]
Comment (nvesely) [reply src/crypto/secretstream/secretstream_macros.rs]: Fixed in https://github.com/sodiumoxide/sodiumoxide/pull/222/commits/4133121eb6294184f6f0e22948bc294989fb9dc5.

File: src/crypto/secretstream/secretstream_macros.rs [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+pub const TAG_REKEY: u8 = $tag_rekey as u8;
+
+/// Tag final
+pub const TAG_FINAL: u8 = $tag_final as u8;
```
[the remark this answers is not in the document]
Comment (nvesely) [reply src/crypto/secretstream/secretstream_macros.rs]: Fixed in https://github.com/sodiumoxide/sodiumoxide/pull/222/commits/4133121eb6294184f6f0e22948bc294989fb9dc5.

File: src/crypto/secretstream/secretstream_macros.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (nvesely) [reply src/crypto/secretstream/secretstream_macros.rs]: https://github.com/sodiumoxide/sodiumoxide/pull/222/commits/4133121eb6294184f6f0e22948bc294989fb9dc5

File: src/crypto/secretstream/secretstream_macros.rs [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+/// `State` contains the state for multi-part (streaming) computations. This allows the caller
+/// to process a message as a sequence of multiple chunks.
+pub struct State($state_name);
+
+/// Initializes a `state` using the `key` and an internal, automatically generated initialization vector.
+/// It then stores the stream header into header (crypto_secretstream_xchacha20poly1305_HEADERBYTES bytes).
+pub fn init_push(key: &Key) -> (State, [u8; HEADERBYTES]) {
+    unsafe {
+        let mut state: $state_name = mem::uninitialized();
+        let mut header: [u8; HEADERBYTES] = mem::uninitialized();
```
[the remark this answers is not in the document]
Comment (nvesely) [reply src/crypto/secretstream/secretstream_macros.rs]: Fixed in https://github.com/sodiumoxide/sodiumoxide/pull/222/commits/4133121eb6294184f6f0e22948bc294989fb9dc5

File: src/crypto/secretstream/secretstream_macros.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (nvesely) [reply src/crypto/secretstream/secretstream_macros.rs]: Fixed in https://github.com/sodiumoxide/sodiumoxide/pull/222/commits/4133121eb6294184f6f0e22948bc294989fb9dc5.

File: src/crypto/secretstream/secretstream_macros.rs [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+        $init_push_name(&mut state, header.as_mut_ptr(), key.0.as_ptr());
+        (State(state), header)
+    }
+}
+
+/// initializes a `state` given a secret `key` and a `header`.
+/// The `key` k will not be required any more for subsequent operations.
+/// It returns Err if the header is invalid.
+pub fn init_pull(header: &[u8; HEADERBYTES], key: &Key) -> Result<State, ()> {
```
[the remark this answers is not in the document]
Comment (nvesely) [reply src/crypto/secretstream/secretstream_macros.rs]: Fixed in https://github.com/sodiumoxide/sodiumoxide/pull/222/commits/4133121eb6294184f6f0e22948bc294989fb9dc5. Now a method of `Decryptor`.

File: src/crypto/secretstream/secretstream_macros.rs [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+        if $init_pull_name(&mut state, header.as_ptr(), key.0.as_ptr()) != 0 {
+            return Err(());
+        }
+        Ok(State(state))
+    }
+}
+
+impl State {
+
+    /// encrypts a message `m` using the `state` and the `tag`.
```
[the remark this answers is not in the document]
Comment (nvesely) [reply src/crypto/secretstream/secretstream_macros.rs]: Fixed in https://github.com/sodiumoxide/sodiumoxide/pull/222/commits/4133121eb6294184f6f0e22948bc294989fb9dc5.

File: src/crypto/secretstream/secretstream_macros.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (nvesely) [reply src/crypto/secretstream/secretstream_macros.rs]: Fixed in https://github.com/sodiumoxide/sodiumoxide/pull/222/commits/4133121eb6294184f6f0e22948bc294989fb9dc5.

File: src/crypto/secretstream/secretstream_macros.rs (hunk 1 above)
Comment (nvesely) [reply src/crypto/secretstream/secretstream_macros.rs]: See https://github.com/sodiumoxide/sodiumoxide/pull/222/files/9b34122b045b9d1ecbf8a806a85e3569d896646f#r202159979. Also note this function is now the method `Decryptor::decrypt`.

File: src/crypto/secretstream/secretstream_macros.rs [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+    }
+    
+    /// function verifies that c (a sequence of bytes) contains a valid ciphertext and authentication tag for the given state state and optional authenticated data ad of length adlen bytes.
+    /// If the ciphertext appears to be invalid, the function returns Err.
+    /// If the authentication tag appears to be correct, the decrypted message is returned with tag.
+    /// Applications will typically call this function in a loop, until a message with the crypto_secretstream_xchacha20poly1305_TAG_FINAL tag is found.
+    pub fn pull(&mut self, c: &[u8], ad: Option<&[u8]>) -> Result<(Vec<u8>, u8),()> {
+        let (ad_p, ad_len) = ad.map(|ad| (ad.as_ptr(), ad.len() as c_ulonglong)).unwrap_or((0 as *const _, 0));
+        let mut m = Vec::with_capacity(c.len() - ABYTES);
+        let mut mlen = m.len() as c_ulonglong;
```
[the remark this answers is not in the document]
Comment (nvesely) [reply src/crypto/secretstream/secretstream_macros.rs]: Fixed in https://github.com/sodiumoxide/sodiumoxide/pull/222/commits/4133121eb6294184f6f0e22948bc294989fb9dc5.

**Comment by Dylan-DPC** [comment]
>  I don't have the permissions to do that because @Dylan-DPC keeps me on a tight leash as a limited-privileges maintainer

@nvesely nothing of that sort. Must be the checkbox.

**Comment by nvesely** [comment]
@Dylan-DPC just giving you a hard time :trollface:

**Comment by Dylan-DPC** [comment]
Superseded by #240

**Pull request closed without merging**